src/Verismith/Verilog2005/AST.hs

```diff
@@ -0,0 +1,1405 @@
+    , _MBParam = _MBParam mba <> _MBParam mbb
+    , _MBLocalParam = _MBLocalParam mba <> _MBLocalParam mbb
+    , _MBDecl = _MBDecl mba <> _MBDecl mbb
+    , _MBSpecParam = _MBSpecParam mba <> _MBSpecParam mbb
+    , _MBBody = _MBBody mba <> _MBBody mbb
+    }
+instance Monoid ModuleBlock where
+  mempty = ModuleBlock [] "" [] [] [] [] [] [] [] Nothing False Nothing (Just NTWire)
+
+-- | Signal level
+data SigLevel = L0 | L1 | LX | LQ | LB
+  deriving (Eq, Ord, Data, Generic, NFData)
+instance Show SigLevel where
+  show x = case x of { L0 -> "0" ; L1 -> "1" ; LX -> "x" ; LQ -> "?" ; LB -> "b" }
+-- | Combinatorial table row
+data CombRow = CombRow { _CRInput :: ![SigLevel], _CROutput :: !ZOX }
+  deriving (Show, Eq, Ord, Data, Generic, NFData)
+-- | Edge specifier
+data Edge
+  = EdgePos_neg !Bool
+  | EdgeDesc
+    { _EDFrom :: !SigLevel
+    , _EDTo   :: !SigLevel
+    }
+  deriving (Eq, Ord, Data, Generic, NFData)
+instance Show Edge where
+  show x = case x of
+    EdgePos_neg b -> if b then "p" else "n"
+    EdgeDesc { _EDFrom = f, _EDTo = t } -> '(' : show f ++ show t ++ ")"
+-- | Seqential table inputs: a list of input levels with at most 1 edge specifier
+data SeqIn
+  = SIComb ![SigLevel]
+  | SISeq ![SigLevel] !Edge ![SigLevel]
+  deriving (Eq, Ord, Data, Generic, NFData)
+instance Show SeqIn where
+  show x = case x of
+    SIComb l -> concatMap show l
+    SISeq l0 e l1 -> concatMap show l0 ++ show e ++ concatMap show l1
+-- | Sequential table row
+data SeqRow = SeqRow
+  { _SRowInput :: !SeqIn
+  , _SRowState :: !SigLevel
+  , _SRowNext  :: !(Maybe ZOX)
+  }
+  deriving (Show, Eq, Ord, Data, Generic, NFData)
+-- | Primitive transition table
+data PrimTable
+  = CombTable ![CombRow]
+  | SeqTable
+    { _STReg  :: !(Either [Attribute] CExpr)
+    , _STInit :: !(Maybe ZOX)
+    , _STRow  :: ![SeqRow]
+    }
+  deriving (Show, Eq, Ord, Data, Generic, NFData)
+-- | Primitive block
+data PrimitiveBlock = PrimitiveBlock
+  { _PBAttr     :: ![Attribute]
+  , _PBIdent    :: !ByteString
+  , _PBOutput   :: !([Attribute], ByteString)
+  , _PBInput    :: ![([Attribute], ByteString)]
+  , _PBBody     :: !PrimTable
+  }
+  deriving (Show, Eq, Ord, Data, Generic, NFData)
+
+-- | Liblist or Use
+data LLU
+  = LLULiblist ![ByteString]
+  | LLUUse
+    { _LLUUIdent  :: !ByteString
+    , _LLUUConfig :: !Bool
+    }
+  deriving (Show, Eq, Ord, Data, Generic, NFData)
+-- | Items in a config block
+data ConfigItem = ConfigItem { _CICell_inst :: !Bool, _CIIdent :: !ByteString, _CILLU :: !LLU }
+  deriving (Show, Eq, Ord, Data, Generic, NFData)
+-- | Config Block: Identifier, Design lines, Configuration items
+data ConfigBlock = ConfigBlock
+  { _CBIdent  :: !ByteString
+  , _CBDesign :: ![ByteString]
+  , _CBBody   :: ![ConfigItem]
+  , _CBDef    :: ![ByteString]
+  }
+  deriving (Show, Eq, Ord, Data, Generic, NFData)
+
+-- | Internal representation of Verilog2005 AST
+data Verilog2005 = Verilog2005
```

Comment:
Additionally, I've noticed that this AST does not support arbitrary annotations like the old AST (i.e. `Verilog2005 a`).  This is probably not necessary, definitely not at first, but I think something like that might be useful to have.  However, the initial purpose for the annotations was for Attribute support, so as this AST already supports that, it may not be necessary.